CGroups management in Kubernetes (#55620)
kubernetes/website: content/en/docs/concepts/architecture/cgroups-management.md

The document concentrates on cgroupv2 and systemd cgroup driver.

<!-- body -->

## Top level Cgroups hierarchy

The root directory for all cgroups in a cgroupv2 system is `/sys/fs/cgroup/`. When using the `systemd` cgroup driver, Kubernetes organizes pods into a hierarchy of systemd slices. The root for all Kubernetes-managed containers is `kubepods.slice`.

So top level hierarchy looks like the following: 

- `/sys/fs/cgroup/`:
  - `kubepods.slice`: Contains all Kubernetes pods.
  - `system.slice`: Contains system services (e.g., `sshd`, `containerd`, `kubelet`).
  - `user.slice`: Contains user sessions.

Kubelet manages cgroup settings of `kubepods.slice` to define it's relative importance to the system workloads. These settings are designed to ensure the node stability.

### Observed Root Cgroup Settings

The following settings were observed on a standard GKE node (`e2-medium`) within `/sys/fs/cgroup/kubepods.slice/` and given as an example. 

Calculated from [Node Allocatable and Reserved Resources](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/):

> - **`cpu.weight`**: `37`. Relative CPU weight for all pods.

Reviewer: I feel like this section could be organized better. maybe broken up by cgroup controller? the order they're described doesn't flow or feel cohesive

Author: I tried to make observations here and next sections are broken down. Please tell if you feel strongly on how I should break it

- **`memory.max`**: `3040550912`. Hard memory limit for the entire `kubepods.slice`.
- **`pids.max`**: `4194304`. Max number of processes allowed across all pods.

The `cpu.weight` defines the amount of CPU shares Kubernetes-managed pods are recieving comparing to the system processes. The default `cpu.weight` for `system.slice` is `100`. See [January 2026 Kubernetes Blog](https://kubernetes.io/blog/2026/01/30/new-cgroup-v1-to-v2-cpu-conversion-formula/) to understand why the shares of `system.slice` are so much higher than `kubepods.slice` in this example.

The `memory.max` and `pids.max` ensures that resources reserved for system daemons will not be consumed by the Kubernetes workloads.

> CPU reservtion for system daemons:

Reviewer: reservation*

Author: fixed

---
title: CGroups management in Kubernetes
content_type: concept
weight: 55
---

<!-- overview -->

This document describes the cgroups management in Kubernetes nodes. This is an overview the describes the approach and is not going into details of all possible customizations.

The document concentrates on cgroupv2 and systemd cgroup driver.

<!-- body -->

## Top level Cgroups hierarchy

The root directory for all cgroups in a cgroupv2 system is `/sys/fs/cgroup/`. When using the `systemd` cgroup driver, Kubernetes organizes pods into a hierarchy of systemd slices. The root for all Kubernetes-managed containers is `kubepods.slice`.

> So top level hierarchy looks like the following: 

Reviewer 2: ```suggestion
A typical top level of slices looks like the following:
```

We shouldn't assume there are no other top level slices.

- `/sys/fs/cgroup/`:
  - `kubepods.slice`: Contains all Kubernetes pods.

>   - `system.slice`: Contains system services (e.g., `sshd`, `containerd`, `kubelet`).

Reviewer 2: nit: we avoid Latin abbreviations. Try "for example".

Author: fixed

  - `user.slice`: Contains user sessions.

> Kubelet manages cgroup settings of `kubepods.slice` to define it's relative importance to the system workloads. These settings are designed to ensure the node stability.

Reviewer 2: How about: 
```suggestion
The kubelet manages the cgroup settings of `kubepods.slice` to define its
relative importance against system services and user sessions.
These slice settings are intended to help assure node stability.
```
?

> ### Observed Root Cgroup Settings

Reviewer 2: I am not sure we need this section. It doesn't particularly fit a reference page. 

If we keep it, put it as the penultimate section.

Author: moved to the bottom

- **`memory.max`**: `3040550912`. Hard memory limit for the entire `kubepods.slice`.
- **`pids.max`**: `4194304`. Max number of processes allowed across all pods.

The `cpu.weight` defines the amount of CPU shares Kubernetes-managed pods are receiving comparing to the system processes. The default `cpu.weight` for `system.slice` is `100`. See [January 2026 Kubernetes Blog](https://kubernetes.io/blog/2026/01/30/new-cgroup-v1-to-v2-cpu-conversion-formula/) to understand why the shares of `system.slice` are so much higher than `kubepods.slice` in this example.

The `memory.max` and `pids.max` ensures that resources reserved for system daemons will not be consumed by the Kubernetes workloads.

CPU reservation for system daemons:

- **`cpuset.cpus`**: `0-1`. Available CPU cores for pods (calculated based on the `ReservedCPUs` or `--reserved-cpus` Kubelet setting).

Default value re-enforcement:

- **`cpu.max`**: `max 100000`. No hard limit (allows using all node CPU).

See [Why not set `cpu.max` on root cgroup?](#why-not-set-cpumax-on-root-cgroup) to learn why it is set to max.

Kubelet can also set other fields, for example hugepages limits.

> ## QoS Slices under `kubepods.slice`

Reviewer 2: nit:
```suggestion
## Child slices for `kubepods.slice` {#kubepods-child-slices}
```
?

    - Path: `/sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod<UID>.slice/`
3.  **BestEffort**: These pods are placed under the `kubepods-besteffort.slice`.
    - Path: `/sys/fs/cgroup/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod<UID>.slice/`

Within each pod slice, there are further sub-groups for each container in the pod.

An example tree for a node with one Guaranteed pod and one Burstable pod:

```
/sys/fs/cgroup/
├── kubepods.slice/
│   ├── kubepods-pod<UID1>.slice/           # Guaranteed pod
│   │   ├── cri-containerd-<ID1>.scope      # container 1
│   │   └── cri-containerd-<ID2>.scope      # container 2
│   ├── kubepods-burstable.slice/
│   │   └── kubepods-burstable-pod<UID2>.slice/  # Burstable pod
│   │       ├── cri-containerd-<ID3>.scope
│   │       └── cri-containerd-<ID4>.scope
│   └── kubepods-besteffort.slice/
├── system.slice/                           # system services
└── user.slice/                             # user sessions
```

The main goal of this hierarchy is to ensure that Pods are getting resources they requested while allowing overcommit of resources.

> ## CPU

Reviewer 2: (nit) I would put "CPU" and "Memory" under a common heading, such as "Infrastructure resource management".

We call them _infrastructure resources_ to distinguish from HTTP resources.

Kubernetes uses two primary mechanisms in cgroupv2 to manage CPU resources:

- **`cpu.weight`**: Controls the proportional share of CPU time. This **replaces** `cpu.shares` from cgroupv1. It ensures that groups receive their requested share of CPU when the node is under load.
- **`cpu.max`**: Sets a hard limit on CPU usage. This **consolidates** the legacy `cpu.cfs_quota_us` and `cpu.cfs_period_us` from cgroupv1 into a single file. It prevents a group from using more than a specific amount of CPU time, even if the node is idle.

These mechanisms work together to provide both resource guarantees and workload isolation.

> ### cgroupv1, cgroupv2 and a new formula

Reviewer 2: Can we consolidate all the comparison with cgroup v1 into a single section of the page? 

Many readers won't need to see how it used to be done.

In cgroupv2, resource control is handled through different files than in cgroupv1. Originally, `cpu.weight` was mapped to `cpu.shares` using the linear formula.

This formula has many issues. For example, under this formula, the standard 1024 shares (1 CPU) maps to approximately **40 weight**. However, systemd's default weight for `system.slice` is **100**. This means that by default, system services are prioritized ~2.5x higher than pods with 1 CPU request.

> So the new non-linear formula is being rolled out as discussed in the [January 2026 Kubernetes Blog](https://kubernetes.io/blog/2026/01/30/new-cgroup-v1-to-v2-cpu-conversion-formula/).

Reviewer 2: We prefer documentation to be timeless; avoid calling it "new".

Author: removed

> ### Principles Kubernetes follows

Reviewer 2: I am not sure about this section either. 

Let's aim to write something that is principally a reference about how the kubelet implements respuy sharing and isolation on Linux nodes. The page ought to be declarative: stating facts and explaining mechanism. We should be overall light on describing the history of Kubernetes and how we got here.

We could however reframe much of what is here under a heading of, perhaps, "Design rationale".

How does that sound?

Author: applied

#### 1. Guaranteed and Burstable Pods
For pods with CPU limits defined:
- **Pod Level**: The `cpu.max` is set at the pod's cgroup slice. It is the sum of the CPU limits of all containers in the pod.
- **Container Level**: Each container also has `cpu.max` set to its specific limit.
- **Example**: A container with a `200m` limit results in `20000 100000` (20ms per 100ms period).

#### 2. BestEffort Pods
BestEffort pods have no CPU limits.
- **Pod and Container Level**: `cpu.max` is set to `max`, meaning no hard limit is enforced.

#### 3. QoS Slices
Kubelet does not typically enforce a hard CPU limit at the QoS tier level.
- **`kubepods-burstable.slice`**: `cpu.max` is set to `max`.
- **`kubepods-besteffort.slice`**: `cpu.max` is set to `max`.

This design allows pods within these slices to burst into any idle CPU cycles on the node, provided they haven't hit their own pod-specific or container-specific limits.

#### 4. Root Cgroup (`kubepods.slice`)
As noted in the responsibilities section, the root `kubepods.slice` also has `cpu.max` set to `max`. This ensures the collective set of all pods can utilize the entire node's CPU capacity if it is not being used by the system.

> ### QoS CPU Weight Calculations

Reviewer 2: ```suggestion
## Algorithm for calculation of CPU weight
```
